client/verta/tests/test_deployment.py

```diff
@@ -579,6 +580,175 @@ def test_dataframe(self, experiment_run, model_for_deployment):
         assert X_train.join(y_train).to_csv(index=False) == six.ensure_str(data_csv)
 
 
+class TestHistogram:
+    @staticmethod
+    def assert_histograms_match_dataframe(histograms, df):
+        """Common assertions for this test suite."""
+        # features match
+        assert set(histograms['features'].keys()) == set(df.columns)
+        # all rows counted
+        assert histograms['total_count'] == len(df.index)
+
+        for feature_name, histogram in histograms['features'].items():
+            series = df[feature_name]
+            histogram_type = histogram['type']
+            histogram_data = histogram['histogram'][histogram_type]
+
+            # all data points counted
+            assert sum(histogram_data['count']) == len(series)
+
+            if histogram_type == "binary":
+                num_false = sum(~series)
+                num_true = sum(series)
+
+                assert histogram_data['count'] == [num_false, num_true]
+            elif histogram_type == "discrete":
+                buckets = histogram_data['bucket_values']
+                counts = histogram_data['count']
+
+                # buckets in ascending order
+                assert buckets == list(sorted(buckets))
+
+                # data within buckets
+                assert all(buckets[0] <= series)
+                assert all(series <= buckets[-1])
+
+                # counts correct
+                for value, count in zip(buckets, counts):
+                    assert sum(series == value) == count
+            elif histogram_type == "float":
+                buckets = histogram_data['bucket_limits']
+                counts = histogram_data['count']
+
+                # buckets in ascending order
+                assert buckets == list(sorted(buckets))
+
+                # data within buckets
+                assert all(buckets[0] <= series)
+                assert all(series <= buckets[-1])
+
+                # counts correct
+                bin_windows = list(zip(buckets[:-1], buckets[1:]))
+                for i, (l, r) in enumerate(bin_windows[:-1]):
+                    assert sum((l <= series) & (series < r)) == counts[i]
+                assert sum(buckets[-2] <= series) == counts[-1]
```

Comment:
This test will pass if `buckets[-1]` is always `1e9` as long as all data is smaller than that. Same for `buckets[0]`. Let's assert we found the right boundaries.

Reply:
I've added checks for the leftmost and rightmost boundaries.

Would it make sense to check the other buckets as well? I would just be copy&pasting the logic from the util, which I think defeats the purpose of testing functionality.

Reply:
No need to check the other ones. Just the cases that are easy to get wrong.

On a side note, you can always verify the intent of the functionality instead of copying the code. For example in this case, we could check 1) are the bucket sizes about the same? 2) do we get the number of buckets we expected? These two are the properties we want. The code in utils is just a way to get to it.